File: StreamTwitt/StreamTwitt/TwitterViewController.m
//
//  ViewController.m
//  StreamTwitt
//
//  Created by Suman Chatterjee on 16/12/2014.
//  Copyright (c) 2014 Suman Chatterjee. All rights reserved.
//

#import "TwitterViewController.h"
#import <Accounts/Accounts.h>
#import <Social/Social.h>
#import "TwitterStreamService.h"
#import "TwittTableViewCell.h"
#import "TwittUtil.h"


@interface TwitterViewController ()<UITableViewDataSource, UITableViewDelegate,TwitterSreamServiceDelegate,UISearchDisplayDelegate, UISearchBarDelegate>

@property (weak, nonatomic) IBOutlet UITableView *twittTableView;
@property (strong, nonatomic) NSMutableArray *dataSource;
@property (strong, nonatomic) UIRefreshControl *refreshControl;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *busyIndecator;
@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
@property (strong, nonatomic)TwitterStreamService *twitterService;
@end

@implementation TwitterViewController

- (void)viewDidLoad {
    [super viewDidLoad];
  //  [self setupRefreshControl]; //Not fully implemented
    self.dataSource = [NSMutableArray array];
    self.busyIndecator.tintColor = [UIColor grayColor];
    
    self.searchBar.delegate = self;
    self.searchBar.placeholder = @"Search for post";
    self.searchBar.tintColor = [UIColor blueColor];
    
    [[UIBarButtonItem appearanceWhenContainedIn:[UISearchBar class], nil] setTitle:@"Search"];
}

- (void)setupRefreshControl
{
    self.refreshControl = [[UIRefreshControl alloc] init];
    self.refreshControl.backgroundColor = [UIColor purpleColor];
    self.refreshControl.tintColor = [UIColor whiteColor];
    [self.refreshControl addTarget:self
                            action:@selector(refresh)
                  forControlEvents:UIControlEventValueChanged];
    [self.twittTableView addSubview:self.refreshControl]; //assumes tableView is @property

}

-(void)refresh {
 //   [self setupTwitterService]; //calls [taskDone] when finished
}

-(void)taskDone {
    [self.refreshControl endRefreshing];
    [self.refreshControl removeFromSuperview];
}
- (void)setupTwitterServiceWithText:(NSString*)string
{
    self.twitterService = [[TwitterStreamService alloc] initWithSearchString:string andNumberOfRequest:10 delegate:self];
    [self.busyIndecator startAnimating];
}

#pragma mark - TwitterSreamService delegate


- (void)twitterFeedAvailable:(NSDictionary *) twitterDict
{
    [self.dataSource addObject:twitterDict];
    if(self.dataSource.count != 0) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.twittTableView reloadData];
        });
    }


}
- (void)twitterFeedLoadFinished
{
    //Twitter loading finished
    [self.busyIndecator stopAnimating];
}

- (void)twitterStreamServiceError
{
    //Handle error
    NSLog(@"Account not found");
    [self.busyIndecator stopAnimating];
    [self taskDone];

}

- (void)noTwitterAccountSetUp
{
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.busyIndecator stopAnimating];
        
        UIAlertController *alert= [UIAlertController
                                   alertControllerWithTitle:@"Twitter"
                                   message:@"Please ge to the settings and set up a Twitter account"
                                   preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * action) {}];
        
        [alert addAction:defaultAction];
        [self presentViewController:alert animated:YES completion:nil];

    });


}

#pragma mark -
#pragma mark UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataSource.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    TwittTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([TwittTableViewCell class]) forIndexPath:indexPath];
    
    NSDictionary * twittData = _dataSource[[indexPath row]];
    [cell configureCellWithText:twittData[@"text"] name:twittData[@"name"] ];
    return cell;
}


#pragma mark - search display controller delegate
- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)searchBar
{
   
    return YES;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    
//    if(!searchText.length){
//        [self checkForRecentSearch];
//        return;
//    }
    
}


- (BOOL)searchBarShouldEndEditing:(UISearchBar *)searchBar
{
    return YES;
}



- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    if(![TwittUtil isEmptyString:searchBar.text]){
        
        [self setupTwitterServiceWithText:searchBar.text];
        
    }
}

- (void)searchWasCancelled
{
}



@end

File: StreamTweetSwift/StreamTweetSwift/TwitterViewController.swift
//
//  ViewController.swift
//  StreamTweetSwift
//
//  Created by Suman Chatterjee on 01/06/2015.
//  Copyright (c) 2015 Suman Chatterjee. All rights reserved.
//

import UIKit

class TwitterViewController:UIViewController,UITableViewDataSource,UITableViewDelegate,TwitterSreamServiceDelegate,UISearchDisplayDelegate,UISearchBarDelegate
{
    
    @IBOutlet weak var twittTableView: UITableView!
    @IBOutlet weak var busyIndecator:UIActivityIndicatorView!
    @IBOutlet weak var twittSearchBar:UISearchBar!
    
    var dataSource: [[String:String]] = []
    var searchActive : Bool = false
    var twitterStreamService :TwitterStreamService!
    
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder:aDecoder)
    }
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        twittTableView.delegate = self
        twittTableView.dataSource = self
        self.busyIndecator.tintColor = UIColor .grayColor()
        self.twittSearchBar.delegate = self
        self.twittSearchBar.placeholder = "Search for Tweets"
        self.twittSearchBar.tintColor = UIColor.blueColor()
    //    self.twittSearchBar.setValue("Search", forKey: "cancelButton")
        
        

    }
    
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    // MARK: UITableViewDataSource
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.dataSource.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        let cellIdentifire = "TwitterViewCell" // NSStringFromClass(TwitterViewCell.self).componentsSeparatedByString(".").last!
        
        
        let cell  = tableView.dequeueReusableCellWithIdentifier(cellIdentifire) as! TwitterViewCell
        let twittDict = dataSource[indexPath.row] as [String : String]
        
        let twittText = twittDict["text"]! as String
        let name = twittDict["name"]! as String
        
        
        cell.configureCellWithText(twittText, name: name)
        return cell
    }
    
    // MARK: UISearchBarDelegate
    func searchBarTextDidBeginEditing(searchBar: UISearchBar) {
        searchActive = true;
        searchBar.showsCancelButton = true
        let uiButton = self.twittSearchBar.valueForKey("cancelButton") as! UIButton
        uiButton.setTitle("Done", forState: UIControlState.Normal)

    }
    
    func searchBarTextDidEndEditing(searchBar: UISearchBar) {
        searchActive = false;
    }
    
    func searchBarCancelButtonClicked(searchBar: UISearchBar) {
        searchActive = false;
        self.setupTwitterServiceWithText(searchBar.text)
    }
    
    func searchBarSearchButtonClicked(searchBar: UISearchBar) {
        searchActive = false;
        self.setupTwitterServiceWithText(searchBar.text)
        
    }
    
    func setupTwitterServiceWithText(textSearch: String)
    {
        if (!textSearch.isEmpty){
        
        if (self.twitterStreamService != nil)
        {
            twitterStreamService = nil
        }
        self.twittSearchBar.resignFirstResponder()
        twitterStreamService  = TwitterStreamService(text:textSearch , numberOfRecords: 10, delegate: self)
        self.busyIndecator.startAnimating()
        }
        else{
            self.showErroMessage("Please enter a search string", messageTitle: "Error")

        }
    }
    
    
    // MARK: TwitterSreamServiceDelegate
        
    func twitterFeedAvailable(twitterData:NSDictionary) {
        println("Here is a feedAvailable")
        self.dataSource.append(twitterData as! [String : String])
        if self.dataSource.count != 0
        {
            dispatch_async(dispatch_get_main_queue(), {
                self.twittTableView.reloadData()
            })
        }
        
    }
    
    func twitterFeedLoadFinished() {
        println("Here is a Finished")
        self.busyIndecator.stopAnimating()
        
        
    }
    func twitterStreamServiceError() {
        println("Here is a Protocal")
        self.busyIndecator.stopAnimating()
        self.showErroMessage("Some error", messageTitle: "Error")

        
    }
    
    func noTwitterAccountSetUp(){
        println("Here is a Protocal")
        
        self.busyIndecator.stopAnimating()
        self.showErroMessage("Please ge to the settings and set up a Twitter account", messageTitle:"Twitter" )
        
    }
    
    func showErroMessage(text : String, messageTitle: String)
    {
        dispatch_async(dispatch_get_main_queue(), {
            let alert = UIAlertController(title: messageTitle, message: text, preferredStyle: .Alert)
            // Create the action.
            let cancelAction = UIAlertAction(title: "Ok", style: .Cancel) { action in
            }
            
            // Add the action.
            alert.addAction(cancelAction)
            self.presentViewController(alert, animated: true, completion: nil)
            
        })

    }
    
    
    // MARK: - Navigation
    
    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        
        if segue.identifier == "twittDetailsIdentifire"{
            if let tweetDescriptionViewController = segue.destinationViewController as? TwitterDetailViewController{
                let index = self.twittTableView.indexPathForSelectedRow()?.row
                let twittDict = dataSource[index!] as [String : String]
                let name = twittDict["name"]! as String
                tweetDescriptionViewController.name = name
            }
            
            
        }
    }


}

